### PR TITLE
attempt to fix #5241 and #5411 inherit from specialized generic typeRel problem

### LINKED ISSUE #5241: inconsistent behavior when calling parent's proc of generic object
the code below works as expected
```nimrod
type
  Fruit = object of RootObj
    color: int
    
  Apple = object of Fruit
   
proc setColor(self: var Fruit, c: int) =
  self.color = c
 
var a: Apple
a.setColor(10) #this one ok

type
  Banana[T] = object of Fruit
    peel: T
    
var b: Banana[int]
b.setColor(11) #this one ok
```

but another similiar code, failed to compile
```nimrod
type
  BaseFruit[T] = object of RootObj
    color: T
    
  Mango = object of BaseFruit[int]
    peel: int
  
  Orange[T] = object of BaseFruit[T]
    peel: T
  
proc setColor[T](self: var BaseFruit[T], c: int) =
  self.color = c
  
var c: Mango
c.setColor(13) #this one failed to compile

var d: Orange[int]
d.setColor(14) # this one is ok. ????
```

the error messages:

```text
parent.nim(34, 2) Error: type mismatch: got (Mango, int literal(13))
but expected one of:
proc setColor(self: var Fruit; c: int)
proc setColor[T](self: var BaseFruit[T]; c: int)
```

this is inconsistent

### PR DESCRIPTION
**Pull request by jangko**

### CODE REVIEW

File: compiler/types.nim [hunk 1]
```diff
@@ -814,6 +815,8 @@ template ifFastObjectTypeCheckFailed(a, b: PType, body: untyped) =
     if tfFromGeneric in a.flags * b.flags and a.sym.id == b.sym.id:
       # ok, we need the expensive structural check
       body
+    elif tfFromGeneric in unionFlags and tfFinal notin unionFlags:
```
Comment (Araq) [inline compiler/types.nim]: I don't understand this. Firstly, why would the types be equal when one comes from a generic instantiation and the other does not. Secondly, what has ``tfFinal`` to do here?

Comment (jangko) [inline compiler/types.nim]: ouch, my terrible mistake, sorry

**Comment by jangko** [comment]
I think I'am on the right path now, but please review carefully

File: compiler/sigmatch.nim [hunk 2]
```diff
@@ -350,6 +350,14 @@ proc handleFloatRange(f, a: PType): TTypeRelation =
       else: result = isIntConv
     else: result = isNone
 
+template genericParamPut(c: var TCandidate; last, fGenericOrigin: PType) =
```
Comment (Araq) [inline compiler/sigmatch.nim]: This should be a proc, if possible. Code bloat is always a concern for rarely executed code.

Comment (jangko) [inline compiler/sigmatch.nim]: no problem

**Comment by Araq** [comment]
I love it. A perfect patch.

**Pull request merged**